[docs] Lit Labs page update (#930)
lit/lit.dev: packages/lit-dev-content/site/docs/libraries/labs.md

</td>
<td>

A plugin for [Eleventy](https://www.11ty.dev) that pre-renders Lit components at build time, with optional hydration.

</td>
<td class="labs-table-links">

[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/eleventy-plugin-lit#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Feleventy-plugin-lit%5D+in%3Atitle "Issues")

</td>
</tr>

<tr>
<td>

[motion](https://www.npmjs.com/package/@lit-labs/motion)

</td>
<td>Animation helpers for Lit templates.</td>
<td class="labs-table-links">

[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/motion#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Fmotion%5D+in%3Atitle "Issues")

</td>
</tr>

<tr>
<td>

[observers](https://www.npmjs.com/package/@lit-labs/observers)

</td>
<td>A set of reactive controllers that facilitate using the platform observer objects.</td>
<td class="labs-table-links">

> [📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/observers#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Fobservers%5D+in%3Atitle "Issues")

Reviewer: ```suggestion
[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/observers#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions/3355 "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Fobservers%5D+in%3Atitle "Issues")
```

</td>
</tr>

<tr>
<td>

[task](https://www.npmjs.com/package/@lit-labs/task)

</td>
<td>A reactive controller for handling asynchronous tasks.</td>
<td class="labs-table-links">

[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/task#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Ftask%5D+in%3Atitle "Issues")

</td>
</tr>

<tr>
<td>

[testing](https://www.npmjs.com/package/@lit-labs/testing)

</td>
<td>A package containing testing utilities for Lit including generated server-side rendered fixtures.</td>
<td class="labs-table-links">

[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/testing#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Ftesting%5D+in%3Atitle "Issues")

</td>
</tr>

<tr>
<td>

[virtualizer](https://www.npmjs.com/package/@lit-labs/virtualizer)

</td>
<td>An element that provides viewport-based virtualization (including virtual scrolling).</td>
<td class="labs-table-links">

> [📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/virtualizer#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Fvirtualizer%5D+in%3Atitle "Issues")

Reviewer 2: ```suggestion
[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/virtualizer#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions/3362 "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+%5Blabs%2Fvirtualizer%5D "Issues")
```

</td>
</tr>

<tr class="subheading"><td colspan=3>Prototyping</td></tr>

<tr>
<td>

[analyzer](https://www.npmjs.com/package/@lit-labs/analyzer)

</td>
<td>A static analyzer for Lit.</td>
<td class="labs-table-links">

[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/analyzer#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Fanalyzer%5D+in%3Atitle "Issues")

</td>
</tr>

<tr>
<td>

[cli](https://www.npmjs.com/package/@lit-labs/cli)

</td>
<td>Command line tool for Lit.</td>
<td class="labs-table-links">

[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/cli#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Fcli%5D+in%3Atitle "Issues")

</td>
</tr>

<tr>
<td>

[router](https://www.npmjs.com/package/@lit-labs/router)

</td>
<td>A component-oriented router API vended as reactive controllers.</td>
<td class="labs-table-links">

> [📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/router#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions "Feedback")<br>[🐞&nbsp;Issues](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Frouter%5D+in%3Atitle "Issues")

Reviewer 2: ```suggestion
[📄&nbsp;Docs](https://github.com/lit/lit/tree/main/packages/labs/router#readme "Docs")<br>[💬&nbsp;Feedback](https://github.com/lit/lit/discussions/3354 "Feedback")<br>[🐞&nbsp;Issues]((https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+%5Blabs%2Frouter%5D "Issues")
```

  order: 3
---

Lit Labs is an umbrella for Lit packages under development that we are actively seeking feedback on. While we encourage real-world use in order to help the feedback process, please note the following important aspects of Labs packages:

- Lit Labs projects are published under the `@lit-labs` npm scope.
- Breaking changes are likely to occur more frequently than in non-labs packages, but they will still respect standard semantic versioning guildelines and all changes will be published to the CHANGELOG files.

> - While bugs in non-lab packages are typically fixed with higher priority, we still treat labs bugs as high priority, and both are triaged as part of the same process.

Reviewer 3: ```suggestion
- While we strive to address all bugs in a timely fashion, bugs in non-labs projects typically receive higher priority than bugs in labs projects.
```

  order: 3
---

> Lit Labs is an umbrella for projects related to Lit that may be experimental or incomplete. Breaking changes are likely to occur more frequently than in non-labs packages, but they will still respect standard semantic versioning guildelines and all changes will be published to the CHANGELOG files. Lit Labs projects are published under the `@lit-labs` npm scope.

Reviewer 4: ```suggestion
Lit Labs is an umbrella for projects related to Lit that may be experimental or incomplete. Breaking changes are likely to occur more frequently than in non-labs packages, but they will still respect standard semantic versioning guidelines and all changes will be published to the CHANGELOG files. Lit Labs projects are published under the `@lit-labs` npm scope.
```

Reviewer 4: I think it might read better (and not sound so scary) if we started out with a more positive tone, and then listed the caveats as a bulleted list so it's not such a wall of text, ala something like:

> Lit Labs is an umbrella for Lit packages under development that we are actively seeking feedback on. While we encourage real-world use in order to help the feedback process, please note the following important aspects of Labs packages:
>
> * Breaking changes are likely to occur more frequently...<snip>
>* When a Lit Labs project is ready to be graduated, it will begin publishing under the `@lit` scope...<snip>
>* We may decide to deprecate a Lit Labs project...<snip>
>
> Feedback is currently being solicited on the following Labs packges:

Author: Good call on making it a bulleted list.

While bugs in non-lab packages are typically fixed with higher priority, we still treat labs bugs as high priority, and both are triaged as part of the same process.

When a Lit Labs project is ready to be graduated, it will begin publishing under the `@lit` scope—for example, the [`@lit/localize`](/docs/localization/overview/) package started out as a Lit Labs project. Once a package graduates, its first version under the `@lit` scope will match that of the latest in `@lit-labs` but only the `@lit` version will receive new updates.

We may decide to deprecate a Lit Labs project. In such cases, we wil notify the community and a deprecation warning will be added to the npm package. It will receive bug fix support for at least 6 months. A record of historical labs packages will be kept on this page.

> Feedback is currently being solicited on the following Labs packges:

Reviewer 4: ```suggestion
Feedback is currently being solicited on the following Labs packages:
```

> | Package | Description | Docs | Feedback |

Reviewer 2: I tried this out here: https://github.com/lit/lit.dev/pull/940

> | Package | Description | Docs | Feedback |

Reviewer 2: Do we want to have the status indicator in this table too? @kevinpschaaf

Reviewer 4: Yeah, like a 🔴 🟠 🟡 🟢 status with a legend maybe?

Reviewer 2: yep

Reviewer 5: sgtm and then maybe sorting them by closest to stable?

Reviewer 5: Status suggestions:

🟢 = Production
🟡 = Evaluation
🟠 = Development
🔴 = At risk (do we need this?)

🟢 react
🟢 ssr
🟢 context
🟢 scoped-registry-mixin
🟡 task
🟡 virtualizer
🟡 motion
🟡 observers
🟡 eleventy-plugin-lit
🟠 router
🟠 testing

Author: Is "Production" the right description? Perhaps that is to say it's close to graduation. I don't think ssr belongs in that category though. That feels more like "Development" based on the potential for API changes.

Reviewer 4: Re:
> 🟢 = Production
> 🟡 = Evaluation
> 🟠 = Development
> 🔴 = At risk (do we need this?)

I don't think "production" makes sense in a labs context at all, and this makes me reconsider whether we even want to try to carve out specific words here with meaning, more than just a spectrum of "stability":

🟢 = Most stable, soliciting final feedback
🟡 = ..
🟠 = ..
🔴 = Least stable, soliciting initial feedback

For example, SSR is _not_ green at this point; there are major critical features like async rendering and event support that could plausibly require disruptive changes. However otoh, it's pretty usable, so for those reasons, I would put it in the "yellow" bucket. But does that mean we're done with "development" and in the "evaluation" phase? No, I don't think so. So I'm not sure how helpful it will be to pin ourselves down to a set of concrete gates like that, rather than just a spectrum of stability.

Reviewer 2: I mostly like this, except I would probably say that 🔴 should be something like "stopped or at risk of being stopped". So green is moving towards graduating and red is the opposite. Packages would start in 🟠  usually.

Reviewer 4: Yeah that makes sense, although a bit nuanced for a legend. Maybe @augustjk can take a crack at what the legend would look like.

Author: I'll try to come up with something. For the labels though, how about this:

🟢 = Near graduation
🟡 = Somewhat stable (?)
🟠 = Soliciting initial feedback
🔴 = At risk of deprecation / Deprecated (?)

Though I'm wondering if the distinction between yellow and orange are really necessary and perhaps we might just combine them to a single "Under active development" ?

Author: I tried splitting up the table into subsections rather than adding an icon to each package and having to label them. WDYT?

<img width="795" alt="image" src="https://user-images.githubusercontent.com/40413829/189777373-fa1b4dd7-f96d-4221-b14a-b4922edfc50e.png">

Open to suggestion on better subheadings.

Author: Preview: https://pr930-5f6cea0---lit-dev-5ftespv5na-uc.a.run.app/docs/libraries/labs/
Styling of subheading is the same as the table here: https://lit.dev/docs/templates/directives/

Reviewer 2: I like this format

> | Package | Description | Docs | Feedback |

Reviewer 2: Looking at the preview, I kind of like it, but I think a monochrome icon would look a lot better than the system emoji: https://pr940-d056ddc---lit-dev-5ftespv5na-uc.a.run.app/docs/libraries/labs/

> | Package | Description | Docs | Feedback |

Reviewer 4: Stacking them in one column seems fine to save horizontal space, but I'd prefer the words to be written out, otherwise I think it's too obtuse:

> [📄 Docs](https://github.com/lit/lit/tree/main/packages/labs/motion#readme)
> [💬 Discussion](https://github.com/lit/lit/discussions)
> [🐞 File Issue](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+%5Blabs%2Fmotion%5D)

> | Package | Description | Docs | Feedback |

Reviewer 2: That works too, but I think reducing the font size for that column would help a lot

> | Package | Description | Docs | Feedback |

Reviewer 2: btw, I don't think "File Issue" is very useful, _unless_ it pre-fills something specific to the labs package. If the link is the same for every package it doesn't deliver much.

> | Package | Description | Docs | Feedback |

Reviewer 4: That was the idea. This fills in the title:

https://github.com/lit/lit/issues/new?template=bug-report.yaml&title=[labs/react]%20%3CBug%20summary%20here%3E

It seems like this might get it to fill in the drop-down based on [this](https://stackoverflow.com/questions/34146618/pre-populate-the-github-new-issue-form-using-the-querystring), but it doesn't work:

https://github.com/lit/lit/issues/new?template=bug-report.yaml&title=[labs/react]%20%3CBug%20summary%20here%3E&affected[]=React%20(@lit-labs/react)

> | Package | Description | Docs | Feedback |

Author: I tried some variations but unfortunately the pre-fill of the form doesn't seem to work for the multiselect, only text fields. :( I'll update the links at least fill the titles though

> | Package | Description | Docs | Feedback |

Author: Actually I like the link being a search with package name better. It would help prevent duplicates and I doesn't presume whether they want to file issue for a bug report or a feature request.

|-|-|-|-|

> | [motion](https://www.npmjs.com/package/@lit-labs/motion) | Animation helpers for Lit templates. This package includes an `animate` directive for adding tweening animations when an element changes state, as well as an `AnimateController` class for programmatically controlling and coordinating animations. You can find a number of motion examples in the [Playground](https://lit.dev/playground/#sample=examples/motion-simple). | [Readme](https://github.com/lit/lit/tree/main/packages/labs/motion#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |

Reviewer 5: ```suggestion
| [motion](https://www.npmjs.com/package/@lit-labs/motion) | Animation helpers for Lit templates. | [Readme](https://github.com/lit/lit/tree/main/packages/labs/motion#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |
```

| [ssr](https://www.npmjs.com/package/@lit-labs/ssr) | A package for server-side rendering Lit templates and components. | [Docs](/docs/ssr/overview)<br>[Readme](https://github.com/lit/lit/tree/main/packages/labs/ssr#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |
| [task](https://www.npmjs.com/package/@lit-labs/task) | A reactive controller for handling asynchronous tasks. | [Readme](https://github.com/lit/lit/tree/main/packages/labs/task#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |
| [virtualizer](https://www.npmjs.com/package/@lit-labs/virtualizer) | An element that provides viewport-based virtualization (including virtual scrolling). | [Readme](https://github.com/lit/lit/tree/main/packages/labs/virtualizer#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |

> | [observers](https://www.npmjs.com/package/@lit-labs/observers) | A set of reactive controllers that facilitate using the platform observer objects, including `MutationObserver`, `ResizeObserver`, `IntersectionObserver`, and `PerformanceObserver`. | [Readme](https://github.com/lit/lit/tree/main/packages/labs/observers#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |

Reviewer 5: ```suggestion
| [observers](https://www.npmjs.com/package/@lit-labs/observers) | A set of reactive controllers that facilitate using the platform observer objects. | [Readme](https://github.com/lit/lit/tree/main/packages/labs/observers#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |
```

> | [context](https://www.npmjs.com/package/@lit-labs/context) | A package containing controllers and decorators for using the [Context Protocol](https://github.com/webcomponents-cg/community-protocols/blob/main/proposals/context.md) as defined by the Web Components Community Group. | [Readme](https://github.com/lit/lit/tree/main/packages/labs/context#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |

Reviewer 5: ```suggestion
| [context](https://www.npmjs.com/package/@lit-labs/context) | A package containing controllers and decorators for using the [Context Protocol](https://github.com/webcomponents-cg/community-protocols/blob/main/proposals/context.md). | [Readme](https://github.com/lit/lit/tree/main/packages/labs/context#readme) | [Discussion](https://github.com/lit/lit/discussions)<br>[File issue](https://github.com/lit/lit/issues/new/choose) |
```
